---
title: "TagSeq code"
author: "K Hunter"
date: "2022-11-02"
output: pdf_document
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Set up the data for reading in edgeR-comes from
<https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4937821.3/>

I think I need to do each individually for ovary and abdomen

```{r}
#set up the environment

library(limma)
library(edgeR)
#read in gene counts with rows being genes/gene counts and columns being samples
files <- list.files(path="/Users/fkhunter12/Downloads/GSE63310_RAW", 
                        pattern=".txt", full.names = TRUE)

#below is just to look at the EntrezID, length, and count for each gene
df<-read.delim(files[1], nrow=5) 

#readDGE is used to (tags=geneID, and last column is counts)
x<-readDGE(files, columns=c(1,3)) 
class(x)
dim(x)

samplenames <- substring(colnames(x), 53, nchar(colnames(x)))

colnames(x) <- samplenames
group <- as.factor(c("LP", "ML", "Basal", "Basal", "ML", "LP", "Basal", "ML", "LP")) 

head(x)
x$samples$group <- group 
lane <- as.factor(rep(c("L004","L006","L008"), c(3,4,2))) 
x$samples$lane <- lane 
x$samples


#Then need to make sure I have file of all the treatment variables associated with text file ID's/beeIDs match
#BiocManager::install("Mus.musculus")
library(Mus.musculus)

geneid <- rownames(x) 
genes <- select(Mus.musculus, keys=geneid, columns=c("SYMBOL", "TXCHROM"), 
                keytype="ENTREZID")
dim(genes)
head(genes)

genes <- genes[!duplicated(genes$ENTREZID),]
genes<-table(geneID, symbol, chromosome) 

x$genes<-genes #Add gene annotations to the DGEList 

x

#need a table with the gene annotations essentially
#should produce this: 
## An object of class "DGEList" 
## $samples 
##                              files group lib.size norm.factors lane 
## 10_6_5_11 GSM1545535_10_6_5_11.txt    LP 32863052            1 L004 
## 9_6_5_11   GSM1545536_9_6_5_11.txt    ML 35335491            1 L004 
## purep53     GSM1545538_purep53.txt Basal 57160817            1 L004 
## JMS8-2       GSM1545539_JMS8-2.txt Basal 51368625            1 L006 
## JMS8-3       GSM1545540_JMS8-3.txt    ML 75795034            1 L006 
## JMS8-4       GSM1545541_JMS8-4.txt    LP 60517657            1 L006 
## JMS8-5       GSM1545542_JMS8-5.txt Basal 55086324            1 L006 
## JMS9-P7c   GSM1545544_JMS9-P7c.txt    ML 21311068            1 L008 
## JMS9-P8c   GSM1545545_JMS9-P8c.txt    LP 19958838            1 L008 
## 
## $counts 
##            Samples 
## Tags        10_6_5_11 9_6_5_11 purep53 JMS8-2 JMS8-3 JMS8-4 JMS8-5 JMS9-P7c JMS9-P8c 
##   497097            1        2     342    526      3      3    535        2        0 
##   100503874         0        0       5      6      0      0      5        0        0 
##   100038431         0        0       0      0      0      0      1        0        0 
##   19888             0        1       0      0     17      2      0        1        0 
##   20671             1        1      76     40     33     14     98       18        8 
## 27174 more rows ... 
## 
## $genes 
##     ENTREZID  SYMBOL  TXCHROM 
## 1     497097    Xkr4     chr1 
## 2  100503874 Gm19938     <NA> 
## 3  100038431 Gm10568     <NA> 
## 4      19888     Rp1     chr1 
## 5      20671   Sox17     chr1 
## 27174 more rows ...
```

Ok! Data is done.

Now onto the conversion of raw counts to counts per million.

It may be important to come back to decide whether this is appropriate
for my data, I suspect so because we are using Tagseq.

```{r}
#these two come from EdgeR
cpm <- cpm(x)
lcpm <- cpm(x, log=TRUE)
#A CPM value of 1 for a gene equates to having 20 counts in the sample with the lowest sequencing depth (JMS9-P8c, library size ≈20 million) or 76 counts in the sample with the greatest sequencing depth (JMS8-3, library size ≈76 million).
#so the log-CPM values are related to the CPM values by log 2(CPM + 2/ average library size)
#For this dataset, the average library size is about 45.5 million, so L ≈ 45.5 and the minimum log-CPM value for each sample becomes log 2(2/45.5) = −4.51. In other words, a count of zero for this data maps to a log-CPM value of −4.51 after adding the prior count or offset:
```

Now, onto the visualization

```{r}
L <- mean(x$samples$lib.size) * 1e-6
M <- median(x$samples$lib.size) * 1e-6
c(L, M)
summary(lcpm)

#check how many genes are lowly expressed #_ =sample number
table(rowSums(x$counts==0)==9) #how many genes have 0 expression across the 9 samples
#In this case it is 5153/22026=23%
        
keep.exprs <- filterByExpr(x, group=group) #filters genes without a worthwhile number of reads-Default is genes with 10 read counts or more in a minmum number of samples 
#For this dataset, the median library size is about 51 million and 10/51 ≈ 0.2, so the filterByExpr function keeps genes that have a CPM of 0.2 or more in at least three samples. 3 because that is the number of replicates....here....
x <- x[keep.exprs,, keep.lib.sizes=FALSE]
dim(x) # 
```

"By default, the function keeps genes with about 10 read counts or more
in a minimum number of samples, where the number of samples is chosen
according to the minimum group sample size. The actual filtering uses
CPM values rather than counts in order to avoid giving preference to
samples with large library sizes. For this dataset, the median library
size is about 51 million and 10/51 *≈* 0.2, so the filterByExpr function
keeps genes that have a CPM of 0.2 or more in at least three samples. A
biologically interesting gene should be expressed in at least three
samples because all the cell type groups have three replicates. The
cutoffs used depend on the sequencing depth and on the experimental
design. If the library sizes had been larger then a lower CPM cutoff
would have been chosen, because larger library sizes provide better
resolution to explore more genes at lower expression levels.
Alternatively, smaller library sizes decrease our ability to explore
marginal genes and hence would have led to a higher CPM cutoff."

```{r}
#Make a figure to look at raw vs. filtered data
lcpm.cutoff <- log2(10/M + 2/L)
library(RColorBrewer)
nsamples <- ncol(x)
col <- brewer.pal(nsamples, "Paired")
par(mfrow=c(1,2))
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col=col, bty="n")
lcpm <- cpm(x, log=TRUE)
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="B. Filtered data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col=col, bty="n")

```

Getting normalization factors

```{r}
x <- calcNormFactors(x, method = "TMM")
x$samples$norm.factors

x2 <- x
x2$samples$norm.factors <- 1
x2$counts[,1] <- ceiling(x2$counts[,1]*0.05)
x2$counts[,2] <- x2$counts[,2]*5
#check how normalized vs unnormalized data is affected 
par(mfrow=c(1,2))
lcpm <- cpm(x2, log=TRUE)
boxplot(lcpm, las=2, col=col, main="")
title(main="A. Example: Unnormalised data", ylab="Log-cpm")
x2 <- calcNormFactors(x2)
x2$samples$norm.factors
## [1] 0.0577 6.0829 1.2202 1.1648 1.1966 1.0466 1.1505 1.2543 1.1090

lcpm <- cpm(x2, log=TRUE)
boxplot(lcpm, las=2, col=col, main="")
title(main="B. Example: Normalised data", ylab="Log-cpm")
```

Yay! Now onto to just doing an MDS plot to see how/whether grouping is
occuring

```{r}
lcpm <- cpm(x, log=TRUE)
par(mfrow=c(1,2))
col.group <- group
levels(col.group) <-  brewer.pal(nlevels(col.group), "Set1")
col.group <- as.character(col.group)
col.lane <- lane
levels(col.lane) <-  brewer.pal(nlevels(col.lane), "Set2")
col.lane <- as.character(col.lane)
plotMDS(lcpm, labels=group, col=col.group) #replace group here with the treatment or thing I want to see if they group together
title(main="A. Sample groups")


plotMDS(lcpm, labels=lane, col=col.lane, dim=c(3,4))
title(main="B. Sequencing lanes")



#OR! check out below!
library(Glimma)
glMDSPlot(lcpm, labels=paste(group, lane, sep="_"), groups=x$samples[,c(2,5)],
         launch=TRUE)
```

**OK, 2.6.23 I stopped here to tackle making an MDS plot!**

Alright, I think this is onto the creating design matrix and contrasts!

##TagSeq data analysis-related to Q2 and Q3 Last edited: 11.2.22

##Q1-What are the molecular signatures of an immune response?

#H1: As stressors impact individuals on a molecular level, a sublethal
immune injection might still influence transcription

#Prediction: Injection and Injection of the e. coli will have an
influence on 1 day individuals

##Visualization

```{r}
#subset to just 1 and 5 day bees
plotPCA(data,
  intgroup = c('e.coli'),
  returnData = FALSE)
#what is the difference here between MDA and PCA plot? 
```

##data analysis

```{r}
#Need to make sure to subset to just controls for each of these
#Also need to figure out how to isolate between each tissue
design <- model.matrix(~0+trt)
design
```

Can use a plot from voom to see whether filtering affects normalization

```{r}
v <- voom(x, design, plot=TRUE)
v
```

Now, need to make the make-contrasts table:

```{r}
contr.matrix<-makeContrasts(ctrl-ringers,ctrl-ecoli,ringers-ecoli,ringers-ecoli, levels=colnames(design))
contr.matrix
```

Then, I will make a linear model

```{r}
vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)
efit <- eBayes(vfit)
plotSA(efit)
topTable(fit)
```

Check out summary below

```{r}
summary(decideTests(efit))

plotMD(tfit, column=1, status=dt[,1], main=colnames(tfit)[1], xlim=c(-8,13))

```

###Q2-What are the molecular signatures of aging?

#H1: As individuals age, their molecular transcription changes.

#Prediction: Control individuals days 1-20 will show changes in
transcription

#Visualization:

-PCA: visualize whether there is overlap between controls of each age.

```{r}
#subset to just controls 
plotPCA(data,
  intgroup = c('Age'),
  returnData = FALSE)
#what is the difference here between MDA and PCA plot? 
```

##Q2.A-Are there signatures of aging in the lab and are they different
than the field?

#Compare across 1, 5, 10, 20 with a simple make_contrasts

#First question is whether there are differences across ages-more of a
means model

```{r}
#Need to make sure to subset to just controls for each of these
#Also need to figure out how to isolate between each tissue
design <- model.matrix(~0+age)
design
```

Can use a plot from voom to see whether filtering affects normalization

```{r}
v <- voom(x, design, plot=TRUE)
v
```

Now, need to make the make-contrasts table:

```{r}
contr.matrix<-makeContrasts(1-5,1-10,1-20,5-10,5-20,10-20, levels=colnames(design))
contr.matrix

```

Then, I will make a linear model

```{r}
agevfit <- lmFit(v, design)
agevfit <- contrasts.fit(agevfit, contrasts=contr.matrix)
ageefit <- eBayes(agevfit)
plotSA(ageefit)
topTable(agefit)
```

Check out summary below

```{r}
summary(decideTests(efit))

plotMD(tfit, column=1, status=dt[,1], main=colnames(tfit)[1], xlim=c(-8,13))

```

To understand whether aging patterns in the lab are similar to what's
seen in other solitary species:

```{r}
#look for list of genes in this paper: https://www.sciencedirect.com/science/article/pii/S0092867414013166 related to fat body aging patterns
# and this website: https://genomics.senescence.info/genes/search.php?organism=Drosophila+melanogaster&show=4

#will need to generate a reciprocal best hit analysis for Nmel to fruit fly (peptide)
#Will need to get the topTable(fit) from the above, potentially as just a 1 vs 20 day individuals. 

#then either use a fisher's exact test or a representation factor (?) to figure out if there is significant overalp 
#Will need to check that the direction is the same 

```

To understand whether aging patterns in the lab are happening and
whether those are comparable to aging patterns of other social insects
species:

##Compare between 1 and 20, then take the top 50 genes that are
downregulated and upregulated and compare to those genes in the
Ti-J-Life pathway (and ESA) pathway (method borrowed from Eusocial
bumblebee queens draft....) (with the same patterns)

```{r}
#To get the genes in the ESA pathways-search this table rstb20190732_si_002 in downloads
#comes from this paper supporting that this pathway is involved in aging in social insects: https://royalsocietypublishing.org/doi/suppl/10.1098/rstb.2019.0732

#To do for the TI-J-LiFE pathway-
```

##Compare to aging in the field

```{r}
#Need to make sure to subset to just the NEs and R 
design_NEvR <- model.matrix(~0+age)
design_NEvR
v_NRvR <- voom(x, design_NEvR, plot=TRUE)
v_NRvR
```

Now, need to do NE vs R comparison to see if the patterns are similar in
the lab

```{r}
contr.matrix_NEvR<-makeContrasts(NE-R, levels=colnames(design_NEvR))
contr.matrix_NEvR
NER_fit <- lmFit(v_NRvR, design_NEvR)
NER_fit <- contrasts.fit(NER_fit, contr.matrix_NEvR)
NER_fit <- eBayes(NER_fit)
plotSA(NER_fit)
topTable(NER_fit)
topTable(agefit)
```

```{r}
#Some type of fisher's exact/representation factor test
#Then need to figure out how to make comparisons in terms of directional shifts 
```

##Q3. From a molecular signature does an immune challenge map onto a
molecular signature of aging? #H1: Because much like aging, an immune
challenge takes a toll-individuals with an immune challenge will be more
like those that have

#Visualization-Use a PCA or venn diagram to see the amount of overlap
across age and between immune challenge treatments

```{r}
#subset to just controls 
plotPCA(data,
  intgroup = c('Age', 'Ecoli'),
  returnData = FALSE)
#what is the difference here between MDA and PCA plot? 

```

(Could also do a venn diagram to see whether there is a higher
similarity between ecoli/ringers and aging-related genes)

#Q3.A-Does this happen in the lab?

#Look at whether there were any effects of the immune challenge (day
1-RvEvC, and 5-RvEvC) then compare those genes to genes differentially
expressed across time (1-20)

#Potentially collapse days 1 and 5 to show overall effects while later
comparing between 1 and 5

#visualize

```{r}
#venn diagram to compare 
```

###statistical tests

```{r}
topTable(agefit) #age top table 
topTable(vfit) #ecoli top table
```

```{r}
#I think I could do a fisher's exact test/calculate representation factor to see the overlap here 
```

```{r}
topTable(NER_fit) #NE vs R
topTable(vfit) #ecoli
```
